### ISSUE TITLE
Rejects-valid with misleading diagnostic for classes defined in variable templates (`template <int> constexpr struct {} test`)

### ISSUE THREAD

**Issue by Eisenwave** [llvm/llvm-project#82499]
```cpp
template <int>
constexpr struct {} test;
```

For this code, clang ouputs:
```
<source>:2:11: error: cannot declare a class template with no name
    2 | constexpr struct {} test;
      |           ^
<source>:2:20: error: expected ';' after struct
    2 | constexpr struct {} test;
      |                    ^
      |                    ;
<source>:2:21: error: a type specifier is required for all declarations
    2 | constexpr struct {} test;
      |                     ^
3 errors generated.
Compiler returned: 1
```

This diagnostic is incorrect. This is not a class template; this is a variable template of unnamed class type, but the class type here is not a template.

**Comment by Eisenwave** [comment]
Nevermind:
> In a [template-declaration](http://eel.is/c++draft/temp.pre#nt:template-declaration), explicit specialization, or explicit instantiation the [init-declarator-list](http://eel.is/c++draft/dcl.decl.general#nt:init-declarator-list) in the declaration shall contain at most one declarator.
When such a declaration is used to declare a class template, no declarator is permitted.

The intent of [[temp.pre] p5](http://eel.is/c++draft/temp.pre#5) seems to be that the declaration is treated as a class template with a declarator, not a variable template.

**Comment by zygoloid** [comment]
Not a rejects-valid, but this is still a bad diagnostic. We should be using the normal heuristics to determine if this is a missing semicolon case or not, and if not, should instead say that a class can't be defined in the type of a variable template.
